feat: Promote NsRecord into a core object

jsantell opened the pull request:
... by extending noosphere_core::data::LinkRecord with its functionality, and removing the noosphere_ns dependency on validating records. Fixes #395

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -65,13 +74,88 @@ pub struct LinkRecord(Jwt);
-            _ => return None,
+    pub fn dereference(&self) -> Option<Cid> {
+        match Ucan::try_from(self) {
+            Ok(ucan) => LinkRecord::link(&ucan),
```
A verb might be nice here e.g., `LinkRecord::find_link` or `extract_link` or something..

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -65,13 +74,88 @@ pub struct LinkRecord(Jwt);
+    pub async fn validate<S: UcanJwtStore>(
+        token: &Ucan,
+        store: &S,
+        opt_did_parser: Option<&mut DidParser>,
```
I've basically given up on including this in signatures. We should just instantiate `DidParser` at the point of use; it's just going to be removed when we update `rs-ucan` for the v0.10.0 UCAN spec.

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -65,13 +74,88 @@ pub struct LinkRecord(Jwt);
+
+    /// Validates the [Ucan] token as a [LinkRecord], ensuring that
+    /// the sphere's owner authorized the publishing of a new
+    /// content address. Returns an `Err` if validation fails.
```
It would probably be good to note the context-specific rules for expiry validation here in this comment.

cdata commented on rust/noosphere-core/src/data/address.rs:
````diff
@@ -99,11 +183,37 @@ impl LinkRecord {
+    /// };
+    /// assert_eq!(LinkRecord::generate_capability(&identity), expected_capability);
+    /// ```
+    pub fn generate_capability(identity: &str) -> Capability<SphereReference, SphereAction> {
````
A generic wrapper for capability generation might be more appropriate e.g., `generate_capability(sphere: &Did, action: SphereAction)`, and we can stick it in `noosphere-core::authority`.

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -65,13 +74,88 @@ pub struct LinkRecord(Jwt);
+    }
+
+    /// The DID key of the sphere that this record maps.
+    pub fn identity(token: &Ucan) -> &str {
```
`sphere_identity`

cdata reviewed (changes requested):
This mostly looks good.

One thing that stood out to me is that _all_ of `LinkRecord`'s implementation is static in order to work around the fact that what it really wants to be operating on is a `Ucan` (but what it has available is an encoded `Jwt`).

It seems like there are two possible revisions to this change that are worth choosing from:

 1. Refactor `LinkRecord` to hold a `Ucan` internally (_not_ a `Jwt`), or....
 2. Add yet-another-wrapper that takes a `LinkRecord` and internally also holds a corresponding `Ucan`
 
After applying one of these approaches, the static methods can be re-imagined as instance methods without extra cost required to parse a `Ucan` for every call.

jsantell commented:
Reconfigured to wrap a Ucan, not dissimilar from NsRecord; updating accordingly

* Promote `noosphere_core::authority::generate_capability(identity: &str, action: SphereAction)`
* Remove optional `DidParser` from `LinkRecord::validate` method
* Updated some traits accordingly, similar to what NsRecord needed as a Ucan wrapper (rather than a Jwt wrapper)
* Updated method names (with no static methods, `dereference` was redundant with `link`, renamed to `get_link()`)

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -99,15 +153,60 @@ impl LinkRecord {
+
+    /// Return a base64-encoded string serialization of the
+    /// underlying [Ucan].
+    pub fn encode(&self) -> Result<String> {
```
I would say just make this `Deref<Target = Ucan>` (which it already is, I think?) and invoke `encode` directly.

Alternatively, it this is important than we should eat our veggies and return a `Result<Jwt>` here.

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -99,15 +153,60 @@ impl LinkRecord {
+    }
+
+    /// Return a base64-encoded [Vec<u8>] of the underlying [Ucan].
+    pub fn as_bytes(&self) -> Result<Vec<u8>> {
```
This seems a little redundant; you can just call `as_bytes`  on any string in Rust

cdata commented on rust/noosphere-core/src/data/address.rs:
```diff
@@ -99,15 +153,60 @@ impl LinkRecord {
+/// [LinkRecord]s compare their [Jwt] representations
+/// for equality. If a record cannot be encoded as such,
+/// they will not be considered equal to any other record.
+impl PartialEq for LinkRecord {
```
This is a pretty expensive equality check. Would you be willing to file an issue for us to implement `PartialEq` for [`Ucan` et al](
https://github.com/ucan-wg/rs-ucan/blob/main/ucan/src/ucan.rs#L20-L42C18) so that we can avoid the encoding cost?

jsantell replied:
Agreed, will do!

cdata reviewed (commented):
Just a little more cleanup requested :pray:

cdata reviewed (approved):
Nice work!
